File: 2022/hs/day-08.hs
import Control.Monad.Trans.State;
import Data.Char (digitToInt)
import Control.Monad
import Data.List (transpose)
import qualified Data.Set as Set

main = do
    input <- readFile "inputs/8.txt"
    let m = matrix input
    let rows = length m
    let cols = length $ head m

    -- part A

    -- I'm leaving my prev solution here just to emphasize, how FUCKING beautiful the new solution is.
    -- Previously I only had the travSeq method, which traverses a sequence to see visible trees.
    -- So my previous solution was to transform the matrix, and use zips to keep track of the coordinates.

    -- let rfwdVisible = fmap (\(i, r) -> zipTrav (fmap (i,) [0..]) r) (zip [0..] m)
    -- let rbwdVisible = fmap (\(i, r) -> zipTrav (fmap (i,) (bwd cols)) r) (zip [0..] (rr m))
    -- let cfwdVisible = fmap (\(j, c) -> zipTrav (fmap (,j) [0..]) c) (zip [0..] (transpose m))
    -- let cbwdVisible = fmap (\(j, c) -> zipTrav (fmap (,j) (bwd rows)) c) (zip [0..] ((rr . transpose) m))

    -- The new solution:
    -- Create an INFINITE MATRIX that consists of all the coordinates.
    -- To create every pass, transform the tree matrix, apply a "forward pass", and untransform the tree matrix
    -- Mark against the coordinate matrix.
    -- THIS IS SUCH A GREAT SOLUTION BECAUSE THE COORDINATE MATRIX STAYS THE SAME!!! LIKE !!!!!!!!!

    let passes = [
            compressMatrix coords . travFr, -- forward row pass
            compressMatrix coords . travBr, -- backward row pass
            compressMatrix coords . travFc, -- forward col pass
            compressMatrix coords . travBc -- backward col pass
            ]

    -- fmap (Set.fromList . ($ m)) passes = for all the passes, apply to m, then convert result into list
    -- foldr Set.union Set.empty $ ... = merge sets via union
    let all = foldr Set.union Set.empty $ fmap (Set.fromList . ($ m)) passes
    print $ Set.size all

    -- part B
    -- Reduce coords to just the matrix length, and flatten
    let coords' = take cols =<< take rows coords

    -- straightforward
    print $ maximum $ fmap (`scenicScore` m) coords'

matrix input = do
    line <- lines input
    return $ fmap digitToInt line

travSeq :: [Int] -> [Bool]
travSeq [] = []
travSeq (t:tx) = True : checkVisible t tx
    where
        checkVisible tallest [] = []
        checkVisible tallest trees = case break (> tallest) trees of
            (hidden, []) -> fmap (const False) hidden
            (hidden, t:tx) -> fmap (const False) hidden ++ True : checkVisible t tx

-- Does a forward row pass through matrix
travFr :: [[Int]] -> [[Bool]]
travFr = fmap travSeq

-- Does a backward row pass through matrix
travBr = revRows . travFr . revRows

-- Does a forward col pass through matrix
travFc = transpose . travFr . transpose

-- Does a backward col pass through matrix
travBc = transpose . travBr . transpose

revRows :: (Functor f) => f [a] -> f [a]
revRows = fmap reverse

compressMatrix :: [[b]] -> [[Bool]] -> [b]
compressMatrix m c = do
    (mr, cr) <- zip m c
    (me, ce) <- zip mr cr
    guard ce
    return me

coords = do
    r <- [0..]
    return $ do
        c <- [0..]
        return (r, c)

-- part B

-- Create axes that indicate all the possible directions one can see from a given tree
-- List will not contain the center node
axes (r, c) m = [
    left,
    right,
    up,
    down
    ]
    where
        row = m !! r
        (left, right) = case splitAt c row of
            sp @ (l, []) -> sp -- shouldn't happen, but y'know safety check
            (l, _:rx) -> (reverse l, rx)

        m' = transpose m
        col = m' !! c
        (up, down) = case splitAt r col of
            sp @ (u, []) -> sp -- shouldn't happen, but y'know safety check
            (u, _:dx) -> (reverse u, dx)

-- Calculate scenic score at a given point
scenicScore (r, c) m = product $ fmap scen (axes (r, c) m)
    where
        tree = m !! r !! c
        -- the number of trees is the number of trees < the height of the center one AND
        -- if there's more trees beyond that, then that the one in front is visible too
        scen s = length visibles + fromEnum ((not . null) other)
            where
                (visibles, other) = span (< tree) s

File: 2022/rs/src/bin/day-08.rs
use std::collections::HashSet;
use std::fs;

fn main() {
    let input = fs::read_to_string("inputs/8.txt").unwrap();
    let m = Matrix::new(&input);
    println!("{:?}", m.visible_from_edge().len());

    let max = (0..m.rows)
        .map(|r| (0..m.cols).map(|c| m.scenic(r, c)).max().unwrap())
        .max().unwrap();
    println!("{max}")
}

struct Matrix {
    m: Vec<Vec<usize>>,
    rows: usize,
    cols: usize
}

struct MatrixCursor<'a> {
    m: &'a Matrix,
    current: Option<(usize /* row */, usize /* col */)>,
    delta: (isize /* row */, isize /* col */)
}

impl Matrix {
    fn new(string: &str) -> Matrix {
        let m = string.lines()
            .map(|line| {
                line.chars()
                    .map(|m| String::from(m).parse().unwrap())
                    .collect()
            })
            .collect();

        let rows = string.lines().count();
        let cols = string.lines().next().unwrap().len();
        Matrix { m, rows, cols }
    }

    fn cursor(&self, at: (usize, usize), delta: (isize, isize)) -> MatrixCursor {
        MatrixCursor { m: self, current: Some(at), delta }
    }

    fn get(&self, row: usize, col: usize) -> Option<usize> {
        self.m.get(row).and_then(|r| r.get(col)).copied()
    }

    fn visible_from_edge(&self) -> HashSet<(usize, usize)> {
        let mut all_visibles = HashSet::new();
        // check rows
        for ri in 0..self.rows {
            let start = (ri, 0);
            let end = (ri, self.cols - 1);
            // down
            let mut tallest = None;
            for (r, c) in self.cursor(start, (0, 1)) {
                let tree = self.get(r, c);
                if tallest < Some(tree) {
                    all_visibles.insert((r, c));
                    tallest.replace(tree);
                }
            }
            // up
            let mut tallest = None;
            for (r, c) in self.cursor(end, (0, -1)) {
                let tree = self.get(r, c);
                if tallest < Some(tree) {
                    all_visibles.insert((r, c));
                    tallest.replace(tree);
                }
            }

            all_visibles.insert(start);
            all_visibles.insert(end);
        }

        // check cols
        for rj in 0..self.cols {
            let start = (0, rj);
            let end = (self.rows - 1, rj);
            // right
            let mut tallest = None;
            for (r, c) in self.cursor(start, (1, 0)) {
                let tree = self.get(r, c);
                if tallest < Some(tree) {
                    all_visibles.insert((r, c));
                    tallest.replace(tree);
                }
            }
            // left
            let mut tallest = None;
            for (r, c) in self.cursor(end, (-1, 0)) {
                let tree = self.get(r, c);
                if tallest < Some(tree) {
                    all_visibles.insert((r, c));
                    tallest.replace(tree);
                }
            }

            all_visibles.insert(start);
            all_visibles.insert(end);
        }

        all_visibles
    }

    fn count_scenic_in(&self, mut it: impl Iterator<Item=(usize, usize)>) -> usize {
        let mut ct = 0;
        if let Some((r, c)) = it.next() {
            let tree = self.get(r, c);
            for (r, c) in it {
                ct += 1;
                if self.get(r, c) >= tree {
                    break;
                }
            }
        }
        ct
    }

    fn scenic(&self, row: usize, col: usize) -> usize {
        [
            self.count_scenic_in(self.cursor((row, col), (0, -1))),
            self.count_scenic_in(self.cursor((row, col), (0, 1))),
            self.count_scenic_in(self.cursor((row, col), (-1, 0))),
            self.count_scenic_in(self.cursor((row, col), (1, 0))),
        ]
            .into_iter()
            .product()
    }
}

impl MatrixCursor<'_> {
    fn advance_cursor(&self, (row, col): (usize, usize)) -> Option<(usize, usize)> {
        #[inline]
        fn checked_add(a: usize, b: isize) -> Option<usize> {
            if b < 0 { 
                a.checked_sub(b.unsigned_abs()) 
            } else {
                a.checked_add(b as usize)
            }
        }

        let (dr, dc) = self.delta;

        let nr = checked_add(row, dr)?;
        let nc = checked_add(col, dc)?;
        
        (nr < self.m.rows && nc < self.m.cols)
            .then_some((nr, nc))
    }
}
impl Iterator for MatrixCursor<'_> {
    type Item = (usize, usize);

    fn next(&mut self) -> Option<Self::Item> {
        let cursor = self.current?;
        self.current = self.advance_cursor(cursor);
        
        Some(cursor)
    }
}